flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBase.java

```diff
@@ -69,17 +87,27 @@ public void onFailure(Throwable t) {
 
 	@Override
 	public void invoke(IN value) throws Exception {
-		if (exception != null) {
-			throw new IOException("invoke() failed", exception);
+		Throwable e = exception.get();
+		if (e != null) {
+			throw new IOException("Error while sending value.", e);
 		}
 		ListenableFuture<V> result = send(value);
+		updatesPending.incrementAndGet();
 		Futures.addCallback(result, callback);
 	}
 
 	public abstract ListenableFuture<V> send(IN value);
 
 	@Override
 	public void close() {
+		while (updatesPending.get() > 0) {
+			synchronized (updatesPending) {
+				try {
+					updatesPending.wait();
+				} catch (InterruptedException e) {
+				}
```

Comment:
Probably better here even: Throw an exception, because closing is incomplete when interrupted (and cannot guarantee correctness)